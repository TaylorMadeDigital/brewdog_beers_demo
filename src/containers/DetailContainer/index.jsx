import React, { useEffect, Fragment } from 'react';
import { useDispatch, useSelector } from "react-redux";
import axios from 'axios';
import { useParams } from "react-router-dom";

import s from './index.scss';
import Heading from '../../components/Heading/';
import Header from '../../components/Header';
import { CONTENT } from '../../consts/content';

const apiUrl = 'https://api.punkapi.com/v2/';

function DetailContainer() {
  const params = useParams();
  const beers = useSelector(state => {
    return state.beers.all
  });
  const current = useSelector(state => {
    return state.beers.current
  });
  const dispatch = useDispatch();
  
  const getList = (items) => {
    return items.map((item, i) => (
      <li key={i}>{item.name}</li>
    ));
  }
  
  useEffect(() => {
    async function fetchBeer() {
      const hasBeer = (beers.find(beer => {
        return beer.id === parseInt(params.id, 10);
      }));
      if(hasBeer) return await dispatch({ type: 'GET_ONE', payload: hasBeer });
      const response = await axios.get(`${apiUrl}beers/${params.id}`);
      await dispatch({ type: 'GET_ONE', payload: response.data[0] });
      return await response.data[0];
    } 
    fetchBeer();
  }, [params.id, beers, dispatch]);

  return (
    <Fragment>
      <Header />
      <article className={s.root}>
        <header className={s.header}>
          <Heading headingLevel="h2">{current.name}</Heading>
          <Heading headingLevel="h3">{current.tagline}</Heading>
          <p>{current.first_brewed}</p>
        </header>
        <div className={s.leftCol}>
          <img src={current.image_url} alt={`${current.name}`} />
        </div>
        <div className={s.rightCol}>
          <p>{current.description}</p>
          <Heading headingLevel="h4" alignment="left">{CONTENT.detail.brewersTips}</Heading>
          <p>{current.brewers_tips}</p>
          <Heading headingLevel="h4" alignment="left">{CONTENT.detail.foodPairing}</Heading>
          <ul>{current.food_pairing && current.food_pairing.map((item, i) => <li key={i}>{item}</li>)}</ul>
          <Heading headingLevel="h4" alignment="left">{CONTENT.detail.ingredients}</Heading>
          <ul>{current.ingredients && getList(current.ingredients.malt)}</ul>
          <ul>{current.ingredients && getList(current.ingredients.hops)}</ul>
        </div>
        <footer>
          {CONTENT.detail.footer} {current.contributed_by}
        </footer>
      </article>
    </Fragment>
  );
}

export default DetailContainer;
