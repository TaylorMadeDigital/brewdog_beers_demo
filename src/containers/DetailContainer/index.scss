@import '../../assets/scss/grid';
@import '../../assets/scss/typography';
@import '../../assets/scss/variables';

.root {
  @include gridContainer;
  padding: 0;
}

.header {
  @include gridItem(12);
  background: $toolbar-bg-color;
  color: $light-text-color;
  padding: $space-base;

  h2, h3 {
    color: $light-text-color;
  }
  h3, p {
    display: inline-block;
  }
  h3 + p {
    margin: 0 $space-s ;
  }
  p {
    @include font-size('small');
  }
}

.leftCol {
  display: none;
  @include breakpoint(medium) {
    display: block;
    @include gridItem(3);
  }

  padding: $space-base;

  img {
    max-width: 100%;
  }
}

.rightCol {
  @include gridItem(12);
  padding: $space-base;
  text-align: left;

  @include breakpoint(medium) {
    @include gridItem(9);
  }
}

footer {
  
  @include font-size('small');
  padding: $space-base;
  text-align: right;
  @include gridItem(12);
}