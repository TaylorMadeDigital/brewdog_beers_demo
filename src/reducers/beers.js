import { BeerActionTypes } from '../actions/index';

const beers = (state = { all: [], current: {}, filter: null}, action) => {
  switch (action.type) {
    case BeerActionTypes.GET_ALL:
      return Object.assign({}, state, {
        all: action.payload
      })
      case BeerActionTypes.GET_ONE:
        return Object.assign({}, state, {
          current: action.payload
        })
    case BeerActionTypes.FILTER_BY:
      return Object.assign({}, state, {
        filter: action.payload
      })
    default:
      return state
  }
}

export default beers;
