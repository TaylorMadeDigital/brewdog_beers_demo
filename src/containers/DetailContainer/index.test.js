import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, cleanup } from "@testing-library/react";
import { createStore } from 'redux'
import { Provider } from 'react-redux';
import DetailContainer from './index';
import beers from './index.fixtures';

describe('<DetailContainer />', () => {

  afterEach(cleanup);

  it('snapshot: renders default state', () => {
    const store = createStore(() => ({ beers: { all: beers, current: beers[3], filter: null }}));
    const { asFragment } = render(<Provider store={store}><Router><DetailContainer /></Router></Provider>);
    expect(asFragment()).toMatchSnapshot();
  });
});
