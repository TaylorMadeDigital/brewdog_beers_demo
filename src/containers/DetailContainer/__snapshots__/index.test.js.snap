// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailContainer /> snapshot: renders default state 1`] = `
<DocumentFragment>
  <a
    href="/"
  >
    <header>
      <img
        alt="Brewdog logo"
        role="presentation"
        src="brewdog_logo.png"
      />
      <h1
        class=""
        data-testid="heading"
      >
        Brewdog Beer Collection
      </h1>
    </header>
  </a>
  <article>
    <header>
      <h2
        class=""
        data-testid="heading"
      >
        Pilsen Lager
      </h2>
      <h3
        class=""
        data-testid="heading"
      >
        Unleash the Yeast Series.
      </h3>
      <p>
        09/2013
      </p>
    </header>
    <div>
      <img
        alt="Pilsen Lager"
        src="https://images.punkapi.com/v2/4.png"
      />
    </div>
    <div>
      <p>
        Our Unleash the Yeast series was an epic experiment into the differences in aroma and flavour provided by switching up your yeast. We brewed up a wort with a light caramel note and some toasty biscuit flavour, and hopped it with Amarillo and Centennial for a citrusy bitterness. Everything else is down to the yeast. Pilsner yeast ferments with no fruity esters or spicy phenols, although it can add a hint of butterscotch.
      </p>
      <h4
        class=""
        data-testid="heading"
      >
        Brewer's Tip
      </h4>
      <p>
        Play around with the fermentation temperature to get the best flavour profile from the individual yeasts.
      </p>
      <h4
        class=""
        data-testid="heading"
      >
        Food Pairing
      </h4>
      <ul>
        <li>
          Spicy crab cakes
        </li>
        <li>
          Spicy cucumber and carrot Thai salad
        </li>
        <li>
          Sweet filled dumplings
        </li>
      </ul>
      <h4
        class=""
        data-testid="heading"
      >
        Ingredients
      </h4>
      <ul>
        <li>
          Extra Pale
        </li>
        <li>
          Caramalt
        </li>
        <li>
          Dark Crystal
        </li>
        <li>
          Munich
        </li>
      </ul>
      <ul>
        <li>
          Centennial
        </li>
        <li>
          Amarillo
        </li>
        <li>
          Centennial
        </li>
        <li>
          Amarillo
        </li>
        <li>
          Centennial
        </li>
        <li>
          Amarillo
        </li>
      </ul>
    </div>
    <footer>
      Contributed by - Ali Skinner &lt;AliSkinner&gt;
    </footer>
  </article>
</DocumentFragment>
`;
